django/recaptcha.md

```diff
@@ -0,0 +1,72 @@
+    - Reference the `.env` file in the project's root `docker-compose.yml` file. This will enable the app to run locally and use your local environment variables. [Here is an example](https://github.com/datamade/parserator.datamade.us/blob/bda3201c3d7873916ed4075a2102b5805fad9a3a/docker-compose.yml#L26).
+    - Add the environment variables to `~/tests/docker-compose.yml`, [like so](https://github.com/datamade/parserator.datamade.us/blob/bda3201c3d7873916ed4075a2102b5805fad9a3a/tests/docker-compose.yml#L13).
+
+Now you can use the public key and private key in your code.
+
+### Implement your code
+There are a few ways of doing this, so [refer to Google](https://developers.google.com/recaptcha/intro) when considering how to do this in your project. For this guide, we are implementing V2 of reCAPTCHA.
+
+1. Configure the view with the public key. [Pass the public key into the template via the context](https://github.com/datamade/parserator.datamade.us/blob/bda3201c3d7873916ed4075a2102b5805fad9a3a/parserator_web/views.py#L519).
+
+2. Add Google's magic JavaScript and public key to your template.
+    -  We have [an example of how this looks](https://github.com/datamade/parserator.datamade.us/blob/bda3201c3d7873916ed4075a2102b5805fad9a3a/parserator_web/templates/parserator_web/sign_up.html#L53) in an application. [See Google's documentation](https://developers.google.com/recaptcha/docs/display#auto_render) for other ways to do this, too.
+
+3. Write a helper function to validate the reCAPTCHA with Google's API. Your backend needs to accept the reCAPTCHA response from the frontend, and then send a request to Google to validate the reCAPTCHA response. 
+    - Here's an example of how we've done this with V2: https://github.com/datamade/parserator.datamade.us/blob/bda3201c3d7873916ed4075a2102b5805fad9a3a/parserator_web/views.py#L524.
+    - Here's another example of this with V3: https://github.com/datamade/django-salsa-auth/commit/c8512d030b90762c7d703bfd1630f79d11e10a5e#diff-bfb393f3e832ecb2f6fb86ad35eefa88a87e0d773b3e4a80ce224d83997df815R137
+
+5. Test it manually and automatically. If everything was setup correctly, then you should be able submit the form. 
+
+By this point, everything should be working in your app so that you can change out the test keys and use a real reCAPTCHA.
+
+## Create a throwaway reCAPTCHA for testing
+If you want to test a real reCAPTCHA, then you'll need to first create one. You can use this reCAPTCHA to test locally and with a review app deployment.
+
+1. Create the reCAPTCHA. Go to the reCAPTCHA admin site and create a new reCAPTCHA: https://www.google.com/recaptcha/admin/create. It should be the v2 type. Add `localhost` as a whitelisted domain.
+
+2. Reconfigure your environment variables. In your `.env` file, change the public and private key. Add the keys from the reCAPTCHA you just created.
+
+3.  Test it. This is where you might need to do some debugging. Good luck!
```

Comment:
I usually use [the dev credentials](https://developers.google.com/recaptcha/docs/faq#id-like-to-run-automated-tests-with-recaptcha.-what-should-i-do), which always pass, instead of creating a throwaway for local development. The docs say it only works for V2, but I used it for V3 without issue in LA Metro.

Reply:
Any advantage to creating a throwaway that I'm missing?

Reply:
[A prior step](https://github.com/datamade/how-to/blob/50d0f175ca09cfee06583fc02dcba4fa3199d958/django/recaptcha.md#set-up-the-test-recaptcha) uses those dev credentials so that a developer can first implement the captcha without any headaches. This step is for testing a "live" captcha locally, just to be thorough. In my mind, this guide is meant to build up to that: first implement in the form with the test captcha, then use a live one to be totally sure it works. I personally wouldn't feel comfortable or confident submitting a PR without testing a real captcha locally. Couldn't think of a better name than "throwaway"...maybe "staging"?

Reply:
Can you say more about how you test a live captcha locally? Is it just confirming that requests go through?

Reply:
I was just creating a real, live captcha and then testing it locally.

Reply:
Could we capture the same benefit by triggering a staging deployment, which should be configured for captcha, and testing, instead of creating a throwaway? This would have the added benefit of confirming we've updated configs correctly, as well.

Reply:
How about instead of a new throwaway for each app, we create Local Development instances for V2 and V3 and recommend using them for local development?

Reply:
This is a good idea. I created these in the captcha admin dashboard and noted this in the documentation. Let me know if that change needs clarity.